### ISSUE TITLE
Champion "Extension function members"

### ISSUE THREAD

**Issue by gafter** [dotnet/csharplang#192]
- [ ] Proposal added
- [ ] Discussed in LDM
- [ ] Decision in LDM
- [ ] Finalized (done, rejected, inactive)
- [ ] Spec'ed

See also https://github.com/dotnet/roslyn/issues/11159

**Comment by Thaina** [comment]
Almost agree and totally support with a bit of arguments and concerns

- Name of this feature make it not sure if this will be able to extend field or not, I would disagree to extend fields
- Extension struct should always pass by reference. Same behaviour as normal struct
- How can we find duplicate operator? When it cause function conflict error it will be hard to find, unlike named member
- Would this syntax will be able to extend `enum` ?
- Is this include extension to implement interface ?

- We don't need new keyword `extension` if we could permit `static class` to extend anything. Reusing syntax

```C#
public static class Ext : MyStruct /// static class cannot extend anything before so it not breaking change
{
}
```

**Comment by Bartmax** [comment]
I think I'm late to the party and forgive my ignorance, but what about just supporting `partial` keyword for `non-partial` classes ?

**Comment by orthoxerox** [comment]
@Bartmax `partial` is for classes in the same assembly. It simply combines several pieces of the same type during compilation. Extensions are for extending arbitrary classes, including those from external compiled assemblies.

**Comment by Bartmax** [comment]
i know that, but that doesn't mean it can be "reworked" to allow classes from other assemblies hence creating extensions for arbitrary classes.

**Comment by alrz** [comment]
I'd suggest the following syntax for extension declarations,
```cs
internal extension class StringExtensions for String { }
internal extension class GenericExtensions<T> for T where T : class { }
```
I believe the name should be optional for convenience (e.g. private nested extension declarations doesn't really need to be named):
```cs
private extension for String { }
private extension<T> for T where T : class { }
```
Similarly, instead of *base list*, we could use `implement` to plug interface/shape/traits to types,
```cs
implement TraitT for String { }
implement<T> FooT for T where T : BarT { }
```
This has various advantages like segregating impls so that they do not show up on *all* instances. Explicit implementations could have the same effect in extension declarations, but I think it's good to separate these concerns (auxiliary methods vs trait implementations).

**Comment by kreig** [comment]
Please consider this feature for the nearest releases. It would make life so much easier.

**Comment by hacklex** [comment]
I think it would be good to have the ability to write extensions for multiple classes in a single context. For example, if you need to cache reflection objects (especially when you're emitting something), sharing said cache would probably be a good idea, and creating an additional static class for that sole purpose would feel somewhat awkward.

Also, I think many would benefit from extensions existing in a non-global context (consider `private static void ExtMethod(this SomeClass x)` being inside a non-static class).

**Comment by paulomorgado** [comment]
@Bartmax, there is no such thing as partial classes. There are partial class definitions and it's a source code feature not an assembly feature (@orthoxerox).

**Comment by gmengano** [comment]
X.0 or 8.0?
[](https://channel9.msdn.com/Blogs/Seth-Juarez/A-Preview-of-C-8-with-Mads-Torgersen#time=32m11s)

**Comment by UweKeim** [comment]
[This is the link to the video](https://channel9.msdn.com/Blogs/Seth-Juarez/A-Preview-of-C-8-with-Mads-Torgersen#time=32m11s), BTW

**Comment by Joe4evr** [comment]
Well, it's 100% in the `X.0 Candidate` milestone. [And you can click it](https://github.com/dotnet/csharplang/milestone/10) if you don't know what that means.

**Comment by IanKemp** [comment]
@Joe4evr That milestone has a release date of January 1 2100 - I think most of us would like to see this feature in a version of C# before we're all dead...

**Comment by JVimes** [comment]
@paulomorgado and @orthoxerox, I'm sure @Bartmax is not suggesting using partial class definitions. He's suggesting using a similar syntax to theirs, but for extension members. It's a great idea:

```
public extension ClassNameHere
{
    // Extension members of all types go here
}
```

**Comment by ivi-hamiti** [comment]
@JVimes I totally agree with your proposed syntax so not to introduce anymore new keyword like the example with the **for** keyword. But what about the cases you are implementing extension methods for types that implement a certain interfaces? Event the naming would be misleading. Like we are doing and extension class for IEnumerable<T> and you have to do something like the following:
```csharp
public extension class IEnumerable<T>
{
    // Extension members of all types go here
    // Don't need to pollute parameter lists with "this ClassName foo"
}
```
Which breaks the naming conventions for C#. Based on your approach i would go with a syntax like this
```csharp
public extension class ExtensionName : ClassName 
{
    // Extension members of all types go here
    // Don't need to pollute parameter lists with "this ClassName foo"
}
```
So it should support the same syntax as class inheritance, generics, etc. The only thing with this approach is that it may be misleading to that user, thinking he should implement the interface or abstract class (as it looks like you are extending/implementing some type.

**Comment by michael-hawker** [comment]
So, would this proposal cover being able to add events as an extension to an existing class as well?

**Comment by jnm2** [comment]
https://github.com/dotnet/roslyn/issues/11159#issue-153598385:

> Limitations:
>     
>  * Events not permitted (at first)

**Comment by leo60228** [comment]
I assume this would support interfaces. This would be really neat, since you could make a mixin like this (with the syntax specified by ivi-hamiti):

```cs
interface IMyMixin {}

public extension class MyMixinImpl : IMyMixin {
    // mixin stuff here
}
```

**Comment by Joe4evr** [comment]
@leo60228 Is there any reason you'd need this over [Default Interface Methods](https://github.com/dotnet/csharplang/blob/master/proposals/default-interface-methods.md) #52 to create mixins?

**Comment by Thaina** [comment]
@Joe4evr - It keep interface being just interface
- It can extend across namespace and assembly
- It more flexible
- It let we have the same expectation to current extension method. Current extension method support interface and generic constraint. Then all object implement extended interface can also call all of its extension method

**Comment by HaloFour** [comment]
@leo60228 

It would support interfaces the same way that extension methods do now.

@Joe4evr @Thaina 

The use of extensions here does not replace default method implementations anymore than extension methods currently do.  They are external, cannot be virtual/specialized and the feature does not enable safe evolution of existing contracts.

**Comment by leo60228** [comment]
No, didn't see that proposal. Sorry!

**Comment by TheFanatr** [comment]
# Type Extensions Syntax Proposition
Please consider the below implied syntax and features.

## Example
This is an example of how using the syntax would look like with the implied syntax and features.

Note: Assume same namespace.

### Dependency
```csharp
public class SampleDependedType
{
    public string SampleStringVariableA { get; protected set; } = String.Empty;

    public string SampleStringVariableB { get; set; } = String.Empty;

    public int SampleIntegerVariableA { get; set; } = 0;
}
```


### Hypothetical Extending Class
```csharp
public extending class SampleExtendingType : SampleDependedType
{
    public string ToString() => $"\{ {SampleStringVariableA}, {SampleStringVariableB}, {SampleIntegerVariableA} \}";

    public void Reset() => (SampleStringVariableA, SampleStringVariableB, SampleIntegerVariableA) = (String.Empty, String.Empty, 0);

    public string SampleStringVariableA { get; set; }
}
```

### Usage of Hypothetical Extending Class
```csharp
using System;

using static SampleExtendingType;

class Program
{
    static void Main()
    {
        var sampleDependedType = new SampleDependedType { SampleStringVariableA = "Hello" };
        var sampleExtendingType = new SampleExtendingType { SampleStringVariableB = "What is up" };

        Console.WriteLine(sampleDependedType.ToString()); // Result: { Hello, , 0 }
        Console.WriteLine(sampleDependedType.SampleIntegerVariableA == sampleExtendingType.SampleIntegerVariableA); // Result: true

        sampleDependedType.Reset();

        sampleExtendingType.SampleIntegerA = 2;

        Console.WriteLine(sampleDependedType.ToString()); // Result: { , , 0 }
        Console.WriteLine(sampleExtendingType.ToString()); // Result: { , What is up, 2 }

        Console.WriteLine(sampleDependedType is sampleExtendingType); // Result: true
    }
}
```

## Usage Explanation
The keyword `extending` would be placed on the type declaration for any class that is to be an extension of another; deriving from a class with the `extending` keyword would cause the new type to extend the derived type. The word `extending` was chosen to reflect the pattern of using present-tense statement word as keywords, such as `protected` and `using`; the `extending` type is the one that is extending the derived class, meaning it is defining an extension to the derived type. Using the extending type statically will cause it to be merged in functionality with the class it is extending, but will still be usable as a standalone type, as a normal derived class.

**Comment by michael-hawker** [comment]
@TheFanatr `protected` and `sealed` are not present-tense words.  `extended` would fit the current pattern of _past tense_ words better.

It'd be also great to see method and operator overload examples in the sample.  Also, a call out that while the sample shows the same namespace, that isn't a restriction/limitation of the feature.

**Comment by TheFanatr** [comment]
> protected and sealed are not present-tense words. extended would fit the current pattern of past tense words better.

Yeah, I know. I think it's a side effect of the fact that I wrote that at 2 AM.

**Comment by sinapis** [comment]
> the keyword to create this syntactical structure should just be extension without any other noun such as class or struct.

Exactly.

**Comment by prad12** [comment]
when is this getting released?

**Comment by LokiMidgard** [comment]
This would be very helpful writing generic wrapper classes that will override the compare Operator only if the generic type argument supports it.

Do I see it correctly that I would need an separate extension class for every concrete generic type. E.g. My Wraper should support `+` when the generic type support it. So I will need one class and not only one Method for each type `int`, `double`, `float`, `long` .....    

```c#
public class E<T>{ /* ...*/ }

public extension class EExtesnions<T>  : E<T>  where T :IComparable<T>
{
   //...
   public static  E<bool> operator <(E<T> left, E<T> right) => // implementation;

   // this is not allowed vvvv
   public static  E<int> operator +(E<int> left, E<int> right) => // implementation;
   public static  E<double> operator +(E<double> left, E<double> right) => // implementation;
}

public extension class EIntExtension  : E<int> 
{
   //...
   public static  E<int> operator +(E<int> left, E<int> right) => // implementation;
}
public extension class EIntExtension  : E<double> 
{
   //...
   public static  E<double> operator +(E<double> left, E<double> right) => // implementation;
}
```

**Comment by theraot** [comment]
Note: I am assuming Language Feature: Extension Everything #11159 is defunt, and this is the replacement.

Any idea how this could be consumed by other languages?

I am not really interested in the syntax, but how these would be compiled and exposed.

---

If we can agree on a nomenclature, I would like to be able to start writing extension members today and have them work - with minimal extra effort - when the new language version rolls out. Not sure if that is possible.

Currently we are able to call extension methods as static methods. I would hope to see something like that for other kinds of members. I am particulary interested in properties.

---

Properties and events could be provided as static methods with a special nomenclature - ie. a method `get_Status` for the getter of an extension property status - and then sprinkle extra syntactic sugar on top (such as the extension keyword suggested above).

Sure, they could use obscure auto-generated names and be made hidden, however that would make it harder for other languages (including old versions of C#) to consume.

So, we could use the auto-generated names for properties, and something similar for events. Similarly, the names generated for operators could be useful too.

---

About implementing interfaces, if that is part of the plan, I suggest the following: To use an object as an interface from an extension, require a cast. Resolve the cast to calling an static method that will wrap the object in a proxy that implements the interface using the other extension members. There is no need to expose the proxy type, only the interface is needed. Also, make sure that method used to do that is available to other languages.

Note: generic interfaces should be possible with generic extension methods.

---

I am not sure about constructors. At least currently there are instances where we use a static `Create` method, thus, I think there is value on extension constructors. I'd suggest an extension method with an `out this` parameter and a particular name (`ctor`?).

Indexers should be possible, if extension properties are ultimately extension methods which are static methods... then they can have more parameters. And for a nameless indexer, it is a matter of picking a name for the methods, so, probably give special treatment `get_Items` and `set_Items`, otherwise `get_this` and `set_this` could work too.

Overload Resolution? If they are just more extension methods, I'd say take the same rules that are in place for extension methods, test it and see if there are problems with that. I guess operators are a special case... If the operands do not have the operator, where to look? The same places you would look for extension methods for the operands.

I would be against providing implicit operators as extension methods. For explicit casting operators, the result could be an out second parameters so that overload resolution can be done based on the type.

I am also against extension fields.

**Comment by AnorZaken** [comment]
So how about keeping it familiar, but still different enough to not cause confusion - instead of the normal single colon : for inheritance, this could use double colons ::
`public static MyExtensions :: ExtendedClass`
Or has anyone suggested this already?

**Comment by PathogenDavid** [comment]
While it isn't commonly used (and wouldn't make sense in that context), it's worth noting that `::` is already in use for [namespace alias qualifiers](https://github.com/dotnet/csharplang/blob/a19fac74c01a6c3da67d38b2f79527145d4edcbc/spec/namespaces.md#namespace-alias-qualifiers).

Personally, I'd rather see the introduction of the `extension` keyword in place of `static` for extension types. It puts more emphasis on the fact that the type plays a special role compared to others.

**Comment by matthew25187** [comment]
Rather than having a special type of extensions class, why not modify the existing syntax for extension methods?  For example,

    public static class DayOfWeekExtensions
    {
      // Methods
      public static bool IsWeekend() extends DayOfWeek
      {
         return this == DayOfWeek.Saturday || this == DayOfWeek.Sunday;
      }
   
      // Properties
      public static bool IsWeekend extends DayOfWeek
      {
         get => this == DayOfWeek.Saturday || this == DayOfWeek.Sunday;
      }
    }

The `extends` keyword defines the context of the `this` keyword within the body of the extension.  This allows for a single extensions class to extend the capabilities of multiple types, as do existing extension classes that currently implement extension methods.  Extension methods could also overload methods from the extended type, so long as the method signature is unique.  If the extended type is later modified to add an overload that is already defined as an extension, the extension overload would have to be flagged as a compilation error.

You could also extend interfaces, where the `this` keyword refers to an instance of the interface.  For example,

    public static class EnumerableExtensions
    {
        public static int CountNulls<T>() extends IEnumerable<T> where T : class
        {
            return this.Count(item => item == null);
        }
    }

When using generics, as in this example, the generic type arguments would be limited to those employed by the extended type so the compiler would know their context at the point the extension was called.

This strategy could also be applied to other constructs such as indexers.

    // Contrived example!
    namespace ExtensionsExample
    {
        public struct FirstName
        {
            private string _name;
    
            public FirstName(string name)
            {
                _name = name;
            }
        }
    
        public struct LastName : IEquatable<LastName>
        {
            private string _name;
    
            public LastName(string name)
            {
                _name = name;
            }
    
            public bool Equals(LastName other)
            {
                return other._name == _name;
            }
        }
    
        public class PersonName
        {
            public FirstName FirstName {get; set;}
            public LastName LastName {get; set;}
        }
    
        public class PersonNameCollection : CollectionBase
        {
            // Collection stuff.
        }
    
        public class Person
        {
            public PersonNameCollection Names {get;}
        }
    
        public static class PersonExtensions
        {
            public static PersonName this[LastName index] extends PersonNameCollection
            {
                get => this.Cast<PersonName>().SingleOrDefault(item => item.LastName.Equals(index));
            }
        }
    }

**Comment by matthew25187** [comment]
How would a class being `sealed` affect the ability of extensions to add capabilities to it?  Should the `sealed` keyword prohibit the definition of extensions on the class?

**Comment by brunoais** [comment]
`sealed` only affects the class. An extension is, in its essence, syntactic sugar for a wrapper around the original class and to make as if the original class had those methods. It never changes that class itself.

**Comment by AnorZaken** [comment]
> When using generics, as in this example, the generic type arguments would be limited to those employed by the extended type so the compiler would know their context at the point the extension was called.

I would never accept the loss of defining my own generic types in an extension.
That's a serious regression of functionality ...or I'm misunderstanding you.

**Comment by willmotil** [comment]
Since you guys are discussing this you may like to see a practical StringBuilder wrapper class that i made and keep updating to that is used in practice.
In order to solve problems with the regular stringbuilder primarily with 2 things that it falls short in 1) it's Insert method 2) c# conversion of numeric variables into strings. Both of these can generate a huge amount of garbage in a realtime situation with massive massive amounts of collections.
Operator Overloading is the least of its problems.

(I know you guys can do it better then me.)

Since you are all pro's here i will not remake a new project just to remove the MonoGame specific methods which will error for straight c# use. They can just be commented out there are only a couple small append methods.

https://github.com/willmotil/MonoGameUtilityClasses/blob/master/MgStringBuilder.cs

The code to illustrate and the output below that in a gif.

```
        MgStringBuilder msg = new MgStringBuilder("Hello World ");
        int counter = 0;
        protected override void QuickDrawSpriteBatch(GameTime gameTime)
        {
            msg.Clear();
            counter++;
            msg = " *operator* "; // fine 
            msg += " Hello " + "World "; // works but creates garbage due to operator overloading limitations.
            //msg.Append(" *Append* ");
            //msg.Append("Hello ").Append("World ").Append(counter); // i can make this work with no garbage.
            
            // print out
            //Console.WriteLine(msg);

            // in my case i draw
            spriteBatch.DrawString(Gu.currentFont, msg, new Vector2(300, 10), Color.Wheat);
        }
```
![gcstringbuilder01](https://user-images.githubusercontent.com/7495196/58372736-e5407100-7eef-11e9-928b-5b542097406e.gif)

**Comment by tullo-x86** [comment]
@willmotil, I think you might have posted your comment on the wrong issue...

**Comment by ddobrev** [comment]
Is there any actual work ongoing? When is this going to be available?

**Comment by Trojaner** [comment]
@ddobrev it is in the X.0 candidate milestone, which means it's on the list for a future version. I assume no work on this is currently done as C# 8 is still not finished.

**Comment by michael-hawker** [comment]
Yeah, I was sad to see previews coming out that didn't have this feature.  Been waiting for this one!

**Comment by lobster2012-user** [comment]
Many of fsharp's features have been ported to csharp. 
Unfortunately, this did not affect custom operators, I hope this will be fixed.
``custom operator |> csharp``

**Comment by ddobrev** [comment]
C# 8 seems finished and this isn't in. When is it going to?

**Comment by CyrusNajmabadi** [comment]
@ddobrev This is no release that this feature is scheduled for.

**Comment by ddobrev** [comment]
But all the syntactic sugar we could safely do without was?
This has been going on for years now. You seem not to grasp that if you keep it up, C# is dead. You pick all the easy sugary tasks and leave anything which requires a little serious work out. One day you'll wake up — most probably when you've lost your pay-check — and realize your competing programming languages are so ahead of you that you can no longer even hope to catch up to them. And all because you've wasted years playing with your toys.

**Comment by CyrusNajmabadi** [comment]
@ddobrev Your comment isn't helpful or constructive.  Please try to stay on topic.  I understand you very much want this feature.  However, this is one of thousands of features that people would like and we're prioritizing based on lots of pieces of data to determine which ones we should ship.

Importantly, many features we do are not at all sugar (though many are as well).  Much of the last few releases has been driving work that is enormously important for scalability and performance (especially for games, server environments, and mobile devices).  

Hopefully in the future you'll find our picks are more what you want.  But that won't always be the case.  Have a good day :)

**Comment by narfanar** [comment]
Any definition doc to read for this proposal? It mentions "Extension Everything" under the roslyn repo. But that's

1. Under a different repo, and is closed.
2. Uses different language from this issue.

What are "Extension function members"? Is there a post missing or is missing a log from an older issue?

**Comment by CyrusNajmabadi** [comment]
@narfanar there is no proposal yet:

![image](https://user-images.githubusercontent.com/4564579/79600070-6b7f1c00-809b-11ea-8e2b-3dcdf2cdee67.png)

**Comment by ChayimFriedman2** [comment]
Personally, I discourage field extensions. This is against the philosophy of adding a helper to the class. If you need to store data, that's not a helper - that data structure. And data structure should be implemented by inheritance, probably.

Also that will face many technical challenges, which will cause the code to be either poor performance or introducing changes in the CLR (or both) because the only ways (do you have others?) to implement this feature is a) implement a dictionary for the field (performing poorly) or b) somehow extends the CLR so objects can have additional fields. But properties (and indexers, etc.) are welcome.

Interfaces are also good (i.e. implementing interface through extensions) since interfaces indicates an **ability** of the type. For instance, be serialized, format itself, etc..

**Comment by NetMage** [comment]
@CyrusNajmabadi Is Mads no longer working?

**Comment by burnss9** [comment]
Hard to determine the current proposal syntax from this thread and I'm not certain where to find it elsewhere. I've seen some mentions of proposals where they keyword static isn't present. I'm fairly uneducated on the backend of how this all works but currently it looks like the IL extension methods are static, if this would go unchanged the `[MethodImpl(MethodImplOptions.Synchronized)]` attribute would be very unintuitive on extension methods/properties. Seems the attribute currently would lock for all instances (makes sense given the static method) but if the static keyword is removed that could be pretty confusing.

**Comment by quinmars** [comment]
The use of `MethodImplOptions.Synchronized` is [discouraged](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.methodimploptions?view=netcore-3.1) anyway:

> Locking on the instance or on the type, as with the Synchronized flag, is not recommended for public types, because code other than your own can take locks on public types and instances. This might cause deadlocks or other synchronization problems.

**Comment by michael-hawker** [comment]
I'm not sure if it'd be part of this proposal or a new feature request. But I just encountered a scenario where I wanted to have an extension that was using a `protected` method, but I'd only be calling the extension method from within the scope of the parent class which has access to the protected method.

It'd be great to be able to flag this type of scenario so the compiler would let me access the protected sub-class methods as long as I'm only using the extension method from a place where I'd have access to those compared to just anywhere.

E.g. In XAML, when extending and writing a custom Control, we inherit from the `Control` class which has a protected `GetTemplateChild` method. However this is generalized and not strongly typed. We wanted to create a typed version which would encapsulate the cast and check for null if we thought it was a required component.

We do this as a private method currently, but it is a general function applicable to anyone inheriting from `Control` and would have been great to generalize and re-use code more effectively with an extension method.

**Comment by PhilParisot** [comment]
I just encountered a scenario where I'm using an API and I'm unable to modify the base class but must use its derived classes (because of added logic existing only in the different derived classes).

I looked into extension methods and it _almost_ gave me what I wanted but I still needed to add custom properties...

Then I looked into adding an interface, and again, it _almost_ did what I needed it to do but interfaces can't tell what class/type it'll be using and therefore can't access said class's public members which the method I was writing again needed.

The method I'd like to use has the exact same implementation on all derived classes, but now I am forced to re-implement it anyways, and it's just not as clean, not to mention time consuming.

Add in the benefit that having the possibility to use class extensions you could write everything you need for your personal use case once, have it work for you, then create a pull request with all the same code (or close) and let the maintainer decide whether to merge or not. In any case your version still works fine for you and it might even get merged and contribute to the ecosystem at large.

It's a win-win, two birds one stone.

**Comment by lukasf** [comment]
Allowing extension methods for implicit conversions would be very helpful to simplify using the new JsonNode APIs in .NET 6:

https://github.com/dotnet/runtime/issues/53195#issuecomment-856223460

I don't believe that implicit casts will ever be added to the actual JsonNode classes. But if we could define them in an extension method in a sub-namespace, users could easily opt-in to implicit conversions by including the sub-namespace.

**Comment by mrahhal** [comment]
I wanted to add one simple to understand scenario in higher level code where this would have been a great feature to have on the library level.

We wanted to support [SQL row value](https://www.sqlite.org/rowvalue.html) syntax in Entity Framework Core (https://github.com/dotnet/efcore/issues/26822). Currently, this is the most readable C# syntax the user will have to write to represent row value:
```cs
.Where(b => EF.Functions.GreaterThan(new[] { b.Column1, b.Column2 }, new[] { 1, 2 }))
```
Not so readable.

With this proposal, we would have been able to overload tuples and allow the user to write the following instead:
```cs
.Where(b => (b.Column1, b.Column2) > (1, 2))
```

**Comment by jl0pd** [comment]
Extension properties would greatly help in [dotnet/comet](https://github.com/dotnet/Comet), which heavily uses builder pattern to create views. 

[Current](https://github.com/dotnet/Comet/blob/513add267636620bfbbd949cd71be45eee94e17e/sample/Comet.Samples/Views/ShapeSample2.cs#L22)
```csharp
new ShapeView(new Circle()
		.Fill(LinearGradient)
		.Style(DrawingStyle.Fill)).Frame(200,100),
```
With extensions
```csharp
new ShapeView(new Circle { Fill = LinearGradient, Style = DrawingStyle.Fill }) { Frame = (200, 100) },
// Frame have signature: void set_Frame(this ShapeView view, (double, double) size)
```

[Current](https://github.com/dotnet/Comet/blob/513add267636620bfbbd949cd71be45eee94e17e/sample/Comet.Samples/Views/SampleApp.cs#L15)
```csharp
new Text("TEST PADDING").Frame(height:30).Margin(top:100),
```
With extensions
```csharp
new Text("TEST PADDING") { FrameHeight = 30, MarginTop = 100, },
```

I find extension property version more readable, especially in first case

**Comment by uecasm** [comment]
Extension static methods would be useful for classes in testing/mocking frameworks and elsewhere that define fluent methods hanging off a static class like `It` or `Is` or `Arg` etc.  Specifically, when you'd like to be able to add an extra method for something custom that works the same way, slightly prettier than making a `MyIt` class or using some fallback that takes a lambda.

**Comment by hawkerm** [comment]
@uecasm was just coming to call out the same thing. Being able to write extensions off of `Assert` for MS Test for example would be useful for folks to write extensions that flow naturally into an existing test framework.

**Comment by NetMage** [comment]
Is there anywhere to see if any progress is being made on this issue?

**Comment by theunrepentantgeek** [comment]
Yup. Right here.

**Comment by uecasm** [comment]
That one's a non-starter for me because it requires an extra keyword/decoration on the class being extended.  The whole point of extension methods is to extend things that the original designer didn't think of and didn't make extensible any other way.

**Comment by LokiMidgard** [comment]
> That one's a non-starter for me because it requires an extra keyword/decoration on the class being extended. The whole point of extension methods is to extend things that the original designer didn't think of and didn't make extensible any other way.

Do you mean #5497? I think it does not…


------------------------------------------------------

*From the proposal*

```c#
public class DataObject
{
    public DataObject this[string member] { get; set; }
    public string AsString() { }
    public IEnumerable<DataObject> AsEnumerable();
    public int ID { get; }
}
```

gets extended by

```c#
public extension JsonDataObject : DataObject
{
    public string ToJson() { … this … }
    public static DataObject FromJson(string json) { … }
}
```

------------------------------------------

I don't see any special keyword. I havn't read the howl thread, so maybe I skiped something

**Comment by uecasm** [comment]
Oh, maybe I overlooked something then.  It looked to me like you could only extend `role`s.

**Comment by LokiMidgard** [comment]
No, as I understood, Roles are a construct to add Interfaces to external types (and maybe more), but are not nessesary to extend something.

**Comment by nathan130200** [comment]
Omg, no ETA for this best feature for C#? 😞

**Comment by brianjlacy** [comment]
What I'm hoping for with extension properties specifically is something like

```csharp
public static class MyStringExtensions {
    public static string Author {
        get(this string);
        set(this string);
    }

    public static string Half {
        get(this string baseStr) => baseStr.Substring(baseStr.Length / 2);
    }
}
```

The examples aren't stellar, but I feel like the syntax feels natural and intuitive given what devs are used to with extension methods.

**Comment by HaloFour** [comment]
@brianjlacy 

Still very much in design, but the current thought is that you declare an "extension type", which can support most members: 

https://github.com/dotnet/csharplang/issues/5497
https://github.com/dotnet/csharplang/blob/main/proposals/extensions.md

```cs
implicit extension MyStringExtensions for string {
    public string Half {
        get => Substring(Length / 2);
    }
}
```

**Comment by nathan130200** [comment]
I still prefer classic extension syntax, but there's some problems:

```cs
public extension MyGuidExtensions for System.Guid // name of qualified name
{
	public string ToBase64()
	{
		// this = treated as Guid instance, internal/private fields should 
		// be visible only in same project or solution or 
		// else use reflection instead.
		// but... "this" can be omitted? this is a good question
		return Convert.ToBase64String(this.ToByteArray());
	}
}

public extension DynIntExtensions for Int32 // force qualified name instead keyword?
{
	public int SevenBitLength
	{
		get
		{
			int numBytes = 0;
			var value = this; // make a copy to self, but how?
			
			do
			{
				value >>>= 7;
				numBytes++;
			}
			while(value > 0);

			return numBytes;
		}
	}
}

// Sample code:

var offset = 0;
int packetId = reader.ReadVarInt32();
var bytesToSkip = packetId.SevenBitLength; 
offset += bytesToSkip

```

**Comment by HaloFour** [comment]
@nathan130200

> // name or qualified name
> // force qualified name instead keyword?

I would expect that you could use the name (or alias) of any type in scope, including the keywords for the primitive types like `int`.

> // this = treated as Guid instance

My gut feeling is that within an "extension" type that `this` would refer to the extended type, e.g. `System.Guid`, and not the extension itself.

> var value = this; // make a copy from self, but how?

If the above is true, then I would expect that this would just work.

These are good questions for the extensions working group.

**Comment by nathan130200** [comment]
@CyrusNajmabadi 

It's not really a problem but the comments I put in the code. Some things don't seem so objective. For example, "this being treated as an instance of Guid".

@HaloFour 

> My gut feeling is that within an "extension" type that this would refer to the extended type, e.g. System.Guid, and not the extension itself.

So in these cases the compiler could understand that if we are inside the extension, it will treat `this` as if it were the instance of the object itself.

Or better perhaps add a keywork specifying `self` that refers to the object we are extending, leaving `this` for the extension itself perhaps.

**Comment by CyrusNajmabadi** [comment]
> "this being treated as an instance of Guid".

I'm not seeing the issue.  Why would that not be hte case?  

> Or better yet, add a keywork called self that refers to the object we are extending, leaving this for the extension itself, 

I don't see the value of that.  if `self` refers to the object being extended... then the current extension still applies to it, since the extension applies to all instances of the extended type.  

>  or can we consider the "extension" type as a reference/value type?

Neither?  The extension is just a set of members on a particular type.  That partial type can be *any* type (unless we choose to limit choices there in some fashion).  

It's like asking "is an extension method a reference or value type?".  It's neither.  It's a method.  An "extension" then is just a collection of those members (all extending a particular type), and it's not limited to just methods.  It can be things like constructors, properties, etc.

**Comment by CyrusNajmabadi** [comment]
As an example, ignoring that your code has a property, you could imagine that:

```
public extension DynIntExtensions for Int32 // force qualified name instead keyword?
{
	public int SevenBitLength
	{
		get
		{
			int numBytes = 0;
			var value = this; // make a copy from self, but how?
			
			do
			{
				value >>>= 7;
				numBytes++;
			}
			while(value > 0);

			return numBytes;
		}
	}
}
```

Is just shorthand for:

```c#
public static class DynIntExtensions 
{
	public int GetSevenBitLength(this int @this)
	{
		int numBytes = 0;
		var value = @this;
			
		do
		{
			value >>>= 7;
			numBytes++;
		}
		while(value > 0);

		return numBytes;
	}
}
```

All that 'implicit extensions' (as a feature) is doing is:

1. allowing you to group tons of related extension members together (instead of having to say what is being extended per member).  
2. allowing you to provide more than just 'extension methods'.  So things like 'extension properties' or 'extension operators' are now all possible.

**Comment by nathan130200** [comment]
What could be done in this case is:

- Support for extension members (fields, methods, events, etc.)
- Allow static types to receive these extensions members as well.

**Comment by CyrusNajmabadi** [comment]
You can't do extension fields as extensions have no way to add state to anything.

**Comment by CyrusNajmabadi** [comment]
> Allow static types to receive these extensions members as well.

They can.

**Comment by CyrusNajmabadi** [comment]
> perhaps an attribute that makes the compiler identify that that extension "method" can be invoked as a member.

That's what an extension already does. See how extension methods are encoded into a dll.  It's just an attribute that marks then as such.

**Comment by CyrusNajmabadi** [comment]
> or something better to identify

We don't need that in source as the source already indicates this by use of the `extension` keyword.

**Comment by weltkante** [comment]
> You can't do extension fields as extensions have no way to add state to anything.

The "workaround" would be an extension property returning a ref-value, which is "almost as good" as a field. It can lookup or create a container based on the instance so it has someplace to store the data, e.g. a ConditionalWeakTable, and then can return a reference into that data container.
